ui/packages/catalog-ui-search/src/main/webapp/js/controllers/cesium.layerCollection.controller.js

```diff
@@ -45,53 +45,64 @@ var Controller = CommonLayerController.extend({
-        model.attributes,
-        'type',
-        'label',
-        'index',
-        'modelCid'
-      )
-
-      if (model.get('type') === 'WMT') {
-        /* If matrixSet is present (OpenLayers WMTS keyword) set tileMatrixSetID (Cesium WMTS keyword) */
-        if (initObj.matrixSet) {
-          initObj.tileMatrixSetID = initObj.matrixSet
-        }
-        /* Set the tiling scheme for WMTS imagery providers that are EPSG:4326 */
-        if (properties.projection === 'EPSG:4326') {
-          initObj.tilingScheme = new Cesium.GeographicTilingScheme()
-        }
+      if (model.get('show')) {
+        this.initLayer(model)
       }
+    }, this)
 
-      var provider = new type(initObj)
+    this.isMapCreated = true
+    return this.map
+  },
+  initLayer: function(model) {
+    const type = imageryProviderTypes[model.get('type')]
+    const initObj = _.omit(
+      model.attributes,
+      'type',
+      'label',
+      'index',
+      'modelCid'
+    )
 
-      /*
-                  Optionally add this provider as a TrustedServer. This sets withCredentials = true
-                  on the XmlHttpRequests for CORS.
-                 */
-      if (model.get('withCredentials')) {
-        const url = require('url')
-        var parsedUrl = url.parse(provider.url)
-        var port = parsedUrl.port
-        if (!port) {
-          port =
-            parsedUrl.protocol === 'https:'
-              ? DEFAULT_HTTPS_PORT
-              : DEFAULT_HTTP_PORT
-        }
-        Cesium.TrustedServers.add(parsedUrl.hostname, port)
+    if (model.get('type') === 'WMT') {
+      /* If matrixSet is present (OpenLayers WMTS keyword) set tileMatrixSetID (Cesium WMTS keyword) */
+      if (initObj.matrixSet) {
+        initObj.tileMatrixSetID = initObj.matrixSet
       }
+      /* Set the tiling scheme for WMTS imagery providers that are EPSG:4326 */
+      if (properties.projection === 'EPSG:4326') {
+        initObj.tilingScheme = new Cesium.GeographicTilingScheme()
+      }
+    }
 
-      var layer = this.map.imageryLayers.addImageryProvider(provider, 0) // the collection is sorted by order, so later things should go at bottom of stack
-      this.layerForCid[model.id] = layer
-      layer.alpha = model.get('alpha')
-      layer.show = model.shouldShowLayer()
-    }, this)
+    const provider = new type(initObj)
 
-    this.isMapCreated = true
-    return this.map
+    /*
+      Optionally add this provider as a TrustedServer. This sets withCredentials = true
+      on the XmlHttpRequests for CORS.
+    */
+    if (model.get('withCredentials')) {
+      const url = require('url')
+      const parsedUrl = url.parse(provider.url)
+      let port = parsedUrl.port
+      if (!port) {
+        port =
+          parsedUrl.protocol === 'https:'
+            ? DEFAULT_HTTPS_PORT
+            : DEFAULT_HTTP_PORT
+      }
+      Cesium.TrustedServers.add(parsedUrl.hostname, port)
+    }
+    const layerId = model.get('id')
+    const layerIndex =
```

Comment:
Okay. Based on my interpretation of the code it seems possible for a layer to be added at the wrong index if the user clicks 'Show' on a layer that was hidden when the map was loaded.

For example, say the map has 4 layers and only the top and bottom layers are configured to be shown when the map is loaded. The maps loads and the user clicks show on the 3rd layer (the second from the bottom). `imageryLayers.length` is 2 and `findIndex()` returns 2 since it's the 3rd layer. So `layerIndex` is 0 and this layer will be added as the bottom-most layer on the map even though it should be the second from the bottom.

Am I understanding this correctly?

Reply:
You're right, good catch. I should probably count the number of shown layers after the index found in this.collection.models to determine the layer placement.